File: FindWeather/FindWeather/src/WeatherFacade.java
import Models.Location;
import Services.LocationService;
import Services.WeatherService;

public class WeatherFacade {
    private final LocationService locationService;
    private final WeatherService weatherService;

    public WeatherFacade() {
        this.locationService = new LocationService();
        this.weatherService = new WeatherService();
    }

    public void getWeatherByIP() {
        try {
            Location location = locationService.getLocationByIP();
            if (location != null) {
                weatherService.fetchWeather(location.getCity(), location.getLatitude(), location.getLongitude());
            } else {
                System.out.println("Could not determine location by IP.");
            }
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }
    }

    public void getWeatherByCity(String city) {
        try {
            Location location = locationService.getLocationByCity(city);
            if (location != null) {
                weatherService.fetchWeather(location.getCity(), location.getLatitude(), location.getLongitude());
            } else {
                System.out.println("Invalid city name or location not found.");
            }
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
}

File: FindWeather/FindWeatherMySolution/facade/WeatherFacade.cs
using FindWeatherMySolution.adopters;
using FindWeatherMySolution.interfaces;
using FindWeatherMySolution.models;
using FindWeatherMySolution.services;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FindWeatherMySolution.facade
{
    public class WeatherFacade
    {
        private readonly ILocationService _locationService;
        private readonly IWeatherService _weatherService;

        public WeatherFacade()
        {
            _locationService = new LocationService();
            _weatherService = new WeatherService(
                new List<IWeatherProvider>
                {
                    new OpenWeatherMapAdapter(),
                    new WeatherStackAdapter()
                }
            );
        }

        public async Task GetWeatherByIp()
        {
            Console.WriteLine("Getting weather data by IP...");

            try
            {
                var location = await _locationService.GetLocationByIp();
                if (location != null)
                {
                    var weatherData = await _weatherService.GetWeather(location);
                    DisplayWeatherData(weatherData);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error: {ex.Message}");
            }
        }

        public async Task GetWeatherByCity(string cityName)
        {
            Console.WriteLine($"Getting weather data for {cityName}...");

            try
            {
                var location = new Location { City = cityName };
                var weatherData = await _weatherService.GetWeather(location);
                DisplayWeatherData(weatherData);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error: {ex.Message}");
            }
        }

        private void DisplayWeatherData(WeatherData weatherData)
        {
            Console.WriteLine($"Data retrieved from {weatherData.Source}.");
            Console.WriteLine($"Weather in {weatherData.Location.City} from {weatherData.Source}: {weatherData.Temperature}°C, {weatherData.Conditions}");
        }
    }
}
